### ISSUE TITLE
HTTP Error 500 with cloudant

### ISSUE THREAD

**Issue by natcohen** [pouchdb/pouchdb#4023]
I sync my application with cloudant and everything worked well until 1-2 days ago. I now get a HTTP error 500; I didn't change anything in the replication process or even in my application. 

Did something change in cloudant to make it incompatible with pouchdb 3.6.0?

**Comment by natcohen** [comment]
The behavior is very strange, I'm also getting the following error: "Uncaught SyntaxError: Unexpected end of input".

If I keep running the replication (with retry set), I'm getting the docs approx one per 5 min with a lot of 500 and Uncaught SyntaxError in between

**Comment by daleharvey** [comment]
Hrm, can you share the database with us to see if we can replicate it, if not can you do a wireshark or something to see what data is actually getting returned to pouchdb?

**Comment by natcohen** [comment]
I can send you my DB, yes... Can you send me your DB link & temporary credentials so I can replicate mine to yours?

**Comment by natcohen** [comment]
@daleharvey I read on internet that this error might come from a malformed JSON in my DB... Any idea how I could quickly check that in all my docs?

**Comment by daleharvey** [comment]
@natcohen hrm, where did you read that? I would be surprised if cloudant let you store invalid json, my email is dale@arandomurl.com, could you send me an email so I can give you credentials, cheers

**Comment by natcohen** [comment]
@daleharvey Did you receive my DB? https://github.com/pouchdb/pouchdb/issues/4040 might be related to that problem...

**Comment by daleharvey** [comment]
@natcohen hrm I didnt get anything in my email, I tried checking spam as well, dale@mozilla.com may work too?

**Comment by natcohen** [comment]
@daleharvey did you receive it this time?

**Comment by daleharvey** [comment]
got it this time, thanks

**Comment by natcohen** [comment]
@daleharvey Did you have a chance to reproduce the error?

**Comment by daleharvey** [comment]
Sorry not yet, its on my list but pretty busy, its sounding like a cloudant issue so trying to debug what actual data is getting returned to pouchdb would be useful if you do have time, cheers

**Comment by corps** [comment]
I can also report that cloudant seems to generating issues as well.  For me, I'm simply getting random requests aborted by the server side.  I don't think it has anything to with pouchdb, although more graceful error handling that @daleharvey landed will be super appreciated!

Talking with cloudant support directly to try to find a root cause.  I would suggest a similar approach.

**Comment by willholley** [comment]
Just to confirm, Cloudant will reject to malformed JSON with status code 400 (Bad Request). A 500 error from Cloudant is unusual and would indicate a bug on our side. @natcohen have you raised a support ticket with Cloudant so we can check the logs at our end?

**Comment by simonaberry** [comment]
@willholley I have also raised a support ticket with Cloudant, but with no real progress.

They did question my security configuration - I had read only access for the pouchdb 'client/user' as I only want a one way replication (cloudant->pouch). They say I need to allow write access or replication access (on the cloudant side)  for it to sync properly  ...?

**Comment by willholley** [comment]
@simonaberry yes, a 403 indicates a permissions problem. As a minimum, you need reader and replication permissions for a pull replication. To push data you need reader and writer.

**Comment by simonaberry** [comment]
OK thanks I give it a try. But it seems weird that 95% of my clients are not having any issues with the current configuration??






On Thu, Jul 16, 2015 at 8:06 AM -0700, "Will Holley" <notifications@github.com> wrote:












@simonaberry yes, a 403 indicates a permissions problem. As a minimum, you need reader and replication permissions for a pull replication. To push data you need reader and writer.

**Comment by willholley** [comment]
@simonaberry I think your issue is that the cookie has expired. Cloudant auth cookies expire after 24 hours so, in response to a 403, you should attempt to re-authenticate using the _session endpoint. I'll lean on others here to advise as to the best way to achieve that in PouchDB but I think it's an orthogonal issue to @natcohen's.

**Comment by daleharvey** [comment]
@willholley how does `replicator` permissions work in cloudant? seems like theres a good chance they will not work well with pouchdb's replicator.

Sync should work fine with readonly permission however, the only issue that could come up is if you delete the remote copy of a read only database

**Comment by willholley** [comment]
@daleharvey `replicator` allows writing of remote checkpoints but nothing else. It allows users to have a "read only" database that can be replicated by others.

**Comment by daleharvey** [comment]
@willholley how does it determine, just writes to `_local/` docs? sounds great though, we can deal with replication of read only databases, however if a user deletes the read only database we cant fallback very well

**Comment by natcohen** [comment]
@willholley unfortunately, I don't have time to work on that until end of next month...if you need my cooperation to resolve an issue, I can give you my DB so you can try to reproduce the problem; this is the best I can do right now. Sorry!

**Comment by willholley** [comment]
@daleharvey y - writing to _local docs and things like _ensure_full_commit.

**Comment by willholley** [comment]
@natcohen just knowing which account/database to look at on the Cloudant end should allow us to determine where the error is (client or server, at least).

**Comment by natcohen** [comment]
@willholley I can't share the DB publicly but I can send it to you. Do you want it?

**Comment by willholley** [comment]
@natcohen I don't need to see the contents of the database - just the Cloudant account name. It's probably easier to do that through the Cloudant support system though (support@cloudant.com or the Cloudant dashboard).

**Comment by natcohen** [comment]
@willholley My DB is completely private so no access to any data without the password... As I said, I don't have enough time to work on that issue until next month.

**Comment by willholley** [comment]
@natcohen Cloudant stores logs of all HTTP requests (minus auth credentials and request bodies). The account name will be enough to correlate a 500 request with our error logs and see if it's a problem on our end. We don't need access to the database contents.

**Comment by natcohen** [comment]
@willholley Thanks, I opened a case and asked for the log file. Hope they won't take too long to answer...

**Comment by willholley** [comment]
@natcohen can you post the error you see in PouchDB (e.g. break on all JS errors and paste the error object here)? PouchDB will return a 500 status code even for errors generated client side (e.g. if the target server is unreachable because the client is offline). As we don't observe the errors on the Cloudant end, a client-side problem seems the most likely thing.

**Comment by natcohen** [comment]
@willholley here is the 500 error... Sometimes I get a serie of "Uncaught SyntaxError: Unexpected end of input"

o {status: 500, name: "unknown_error", message: "Database encountered an unknown error", error: true, result: Object}
error: true
message: "Database encountered an unknown error"
name: "unknown_error"
result: Object
doc_write_failures: 1
docs_read: 1
docs_written: 0
end_time: Thu Jul 23 2015 16:45:54 GMT+0200 (Europe de l’Ouest (heure d’été))
errors: Array[1]
0: o
error: true
message: "Database encountered an unknown error"
name: "unknown_error"
result: Object
doc_write_failures: 1
docs_read: 1
docs_written: 0
end_time: Thu Jul 23 2015 16:45:54 GMT+0200 (Europe de l’Ouest (heure d’été))
errors: Array[1]
last_seq: 965
ok: false
start_time: Thu Jul 23 2015 16:45:43 GMT+0200 (Europe de l’Ouest (heure d’été))
status: "aborting"
__proto__: Object
status: 500
__proto__: r
length: 1
__proto__: Array[0]
last_seq: 965
ok: false
start_time: Thu Jul 23 2015 16:45:43 GMT+0200 (Europe de l’Ouest (heure d’été))
__proto__: Invalid Date
status: "aborting"
__proto__: Object
status: 500__proto__: rconstructor: r(e){Error.call(e.reason),this.status=e.status,this.name=e.error,this.message=e.reason,this.error=!0}toString: (){return JSON.stringify({status:this.status,name:this.name,message:this.message})}
__proto__: d

**Comment by natcohen** [comment]
@willholley Any idea of what it could be? What can I do to move forward?

**Comment by daleharvey** [comment]
We have started seeing a very similiar failure in our test suite as a constant regression, introduced sometime yesterday

https://travis-ci.org/pouchdb/pouchdb/jobs/72304282

How aligned is cloudant and couch-master these days? is this something that could have landed in cloudant then on couch-master, either way we have a failing test so hopefully it related to this bug, probably wont get time tonight but next time I have time will take a look

**Comment by willholley** [comment]
@daleharvey interesting. @rnewson is probably best placed to comment around which recent Cloudant changes have been pushed to CouchDB master as well.

**Comment by rnewson** [comment]
cloudant and couch-master are not aligned today, but that is coming soon. We intend to switch to a couchdb 2.0 base in the next month (all appendages crossed).

An error encountered midway through a response will result in an abrupt socket closure as neither HTTP nor the couchdb/cloudant API has an out-of-band error mechanism. We return errors if we encounter them before we start responding (and communicate that through the status code and response body), but once we've begun the response we have no real choice but to kill the connection and rely on clients, such as the one above, to notice a partial response as an error state.

On our side, we will have logged the error we would have transmitted to the user if we had the ability to do so, so timestamps would be useful.

**Comment by simonaberry** [comment]
this may well be unrelated, but my issue #4040 which @daleharvey cross referenced to this issue turned out to be the same as #4092 : i.e. i was hitting the 10 second timeout on a replication (with no 'live' or 'retry' options). Implementing @nolanlawson 's suggested algorithm to grow the timeout solved my issue.

**Comment by willholley** [comment]
@daleharvey In my local test environment,https://github.com/apache/couchdb-chttpd/commit/57427cf103e9b3f77a715acf2298b2a55b5bd2a3 breaks test "[#2858 {binary: true} in live changes, mixed atts](https://github.com/pouchdb/pouchdb/blob/master/tests/integration/test.attachments.js#L496)" with an "Unexpected end of input" error. I'll dig into it further...

**Comment by natcohen** [comment]
@daleharvey @willholley Any news on that one?

**Comment by gabrielmancini** [comment]
i had this error too.
when i try sync my dbs this error happens
{"error":"function_clause","reason":null,"ref":2864846897}

**Comment by willholley** [comment]
@gabrielmancini that sounds like another orthogonal issue. Can you raise a Cloudant support ticket with the details please?

**Comment by daleharvey** [comment]
@natcohen can you resend me the database please, cant find it in my email archives

**Comment by gabrielmancini** [comment]
the cloudant ticket says:

The error occurs because the document EA080D90-E947-9003-99CA-B2FB2C6CDE50 doesn't exist. The new_edits:false field should only be specified for existing documents.

This looks like a possible PouchDB bug so I suggest you open a GitHub issue in the PouchDB project with the details.

@daleharvey

**Comment by willholley** [comment]
@natcohen @daleharvey - sorry ignore that. I think it's a problem on the Cloudant end.

**Comment by willholley** [comment]
@gabrielmancini it looks like you have a validate doc update function which is throwing an exception. CouchDB / Cloudant could provide a more useful error message in this situation but deleting / fixing the VDU should allow replication to work.

**Comment by gabrielmancini** [comment]
<o> Oh!! i am sorry, u are right!! what a shame... @willholley thanks, when i remove de validation the app works like a charm...

**Comment by rnewson** [comment]
a vdu is allowed to throw {forbidden: "text here"} and {unauthorized:"text here"}. you can also throw("string here") and it'll report it as unknown_error, but anything else is a crash, couchdb doesn't handle it at all well, it percolates up in an ugly fashion, ultimately it's a function_clause.

So this is better error handling for bad input than a bug as such.

**Comment by daleharvey** [comment]
@natcohen this is now very high on my list of things to look at, any chance you could send me the couch file again or give any other help reproducing the error?

**Comment by natcohen** [comment]
@daleharvey Hey, sorry for the delay, I'm quite busy these days... did you receive the DB? I didn't test it since I posted the issue.

**Comment by daleharvey** [comment]
So I tested the database locally and have been able to replicate fine, I dont think there has been any changes that should cause any incompatibility so I think this is possibly a cloudant service issue, as said 500 responses generally mean an unexpected error on the server. 

Since this issue has got a few things confused in it going to close it out, if you are still seeing it it would be great if you could try to isolate what is happening and report as a new issue, thanks a lot
